test/framework/clusterctl/e2e_config_test.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterctl
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	"sigs.k8s.io/cluster-api/internal/goproxy"
+)
+
+func Test_resolveReleaseMarker(t *testing.T) {
+	clientGoproxy, muxGoproxy, teardownGoproxy := newFakeGoproxy()
+	defer teardownGoproxy()
+
+	// setup an handler with fake releases
+	muxGoproxy.HandleFunc("/github.com/o/r1/@v/list", func(w http.ResponseWriter, r *http.Request) {
+		testMethod(t, r, "GET")
+		fmt.Fprint(w, "v1.2.0\n")
+		fmt.Fprint(w, "v1.2.1-rc.0\n")
+	})
+	tests := []struct {
+		name          string
+		releaseMarker string
+		want          string
+		wantErr       bool
+	}{
+		{
+			"Invalid url",
+			"github.com/o/doesntexist",
+			"",
+			true,
+		},
+		{
+			"Get stable release",
+			"go://github.com/o/r1@v1.2",
+			"1.2.0",
+			false,
+		},
+		{
+			"Get latest release",
+			"go://github.com/o/r1@latest-v1.2",
+			"1.2.1-rc.0",
+			false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			got, err := resolveReleaseMarker(tt.releaseMarker, clientGoproxy)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).ToNot(HaveOccurred())
+
+			g.Expect(got).To(BeEquivalentTo(tt.want))
+		})
+	}
+}
+
+func testMethod(t *testing.T, r *http.Request, want string) {
+	t.Helper()
+	if got := r.Method; got != want {
+		t.Errorf("Request method: %v, want %v", got, want)
+	}
+}
+
+// newFakeGoproxy sets up a test HTTP server along with a github.Client that is
+// configured to talk to that test server. Tests should register handlers on
+// mux which provide mock responses for the API method being tested.
+func newFakeGoproxy() (client *goproxy.Client, mux *http.ServeMux, teardown func()) {
```

Comment:
We should think about moving these to a package and use it from there for all occurencies.

As of now this duplicates code from https://github.com/kubernetes-sigs/cluster-api/blob/06e5737724cbc15e299818d59a35b90a93afcbc4/test/framework/clusterctl/e2e_config_test.go .

Reply:
yes I agree, do you have any suggestions on where we can add it? right now it is defined in three places:

- cluster-api/cmd/clusterctl/client/repository/repository_github_test.go
- cluster-api/internal/goproxy/goproxy_test.go
- cluster-api/test/framework/clusterctl/e2e_config_test.go

Reply:
> `NewFakeGoproxy` is already exported inside `cluster-api/internal/goproxy/goproxy_test.go`.
> Maybe we can do the same with the testMethod function and use it from there?

@chrischdi `testMethod` is also already in the same [file](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/goproxy/goproxy_test.go#L204-L210), why not to use them from there?

Reply:
That's exactly what I proposed :-)

Reply:
> That's exactly what I proposed :-)

ah sorry, then i misunderstood your comment 🙂

Reply:
fixed, moved `NewFakeGoproxy` and `TestMethod` to `goproxy.go` and removed duplication in `e2e_config_test.go` and `repository_github_test.go` , please take a look

Reply:
fixed